enhancements/oc/must-gather-collection-mode.md

```diff
@@ -0,0 +1,207 @@
+---
+title: must-gather-collection-mode
+authors:
+  - "@ardaguclu"
+reviewers:
+  - "@ingvagabund"
+approvers: 
+  - "@deads2k"
+api-approvers: 
+  - "@deads2k"
+creation-date: 2024-11-22
+last-updated: 2024-11-22
+tracking-link: 
+  - https://issues.redhat.com/browse/OCPBUGS-37344
+see-also:
+  - "/enhancements/oc/must-gather.md"
+  - "/enhancements/oc/inspect.md"
+---
+
+# must-gather: Collection Mode
+
+## Summary
+
+This proposal introduces a new flag (i.e. `--collection-mode`) in must-gather targeting to large clusters by skipping some
+logs to take less time and storage size. In addition to that, this proposal introduces a new `--node-selector` flag
+in oc adm inspect command to only collect the daemonset pods running on the given node selector.
+
+## Motivation
+
+must-gather, due to its nature, aims to collect every log in the cluster in best effort to provide extensive insights.
+However, this comes with a drawback that on large clusters (e.g. clusters whose node count is greater than 20), completion
+of the must-gather takes excessive time and storage which eventually hurts the usability (or worse, collection failure). As maintainers, we are usually
+under the pressure of two opposite sides; adding more and more logs for better troubleshooting experience, cutting some logs
+for short completion duration and less storage size. It is hard to find the optimum balance in regard to the default behavior.
+
+As a result, we need to find a mitigation plan for large clusters by skipping some logs that are marked as less critical
+meanwhile preserving the default behavior and expectations.
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator maintaining 250 nodes of cluster, I want to have a mechanism to collect the logs quickly and efficiently. 
+Besides, I'm fine skipping some logs and providing them separately. Because currently must-gather takes 6 hours that even may end up with a failure.
+
+#### Story 2
+
+As a cluster administrator maintaining 250 nodes of clusters. I usually need to troubleshoot networking
+that are directly associated to the daemonset pod logs running on workers nodes. So that, I would like to collect everything without skipping
+any log (accepting the long time and a risk of collection failure).
+
+### Goals
+
+1. Introducing a new `--collection-mode` in must-gather that will be used on large clusters.
+2. Introducing a new `--node-selector` in inspect command that will be used to collect daemonsets logs only on the given node selector.
+
+### Non-Goals
+
+1. Changing the default behavior of must-gather and inspect.
+
+## Proposal
+
+### `oc adm must-gather`
+
+There is a new flag in must-gather, namely `--collection-mode`. This flag's default value is set to `medium` that
```

Comment:
It is important to define what the current behavior is. Looking a year back at the proposal from now will give "the current behavior" a different meaning. The same for the limited mode. Worth making a survey to identify what resources we currently collect and categorizing them based on their importance and impact on solving customer issues/debugging clusters.

Reply:
Maybe the wording should change "the current behavior" to something that makes it clearer that it is the behavior before this feature was introduced?

In what regards the survey, I agree and believe it is a good idea. But I guess that the impact on solving issues shouldn't be the only driver, but other factors like potential collection size may matter.

Reply:
+1 for identifying a range of sizes of collected data. Some scripts are collecting more or less constant or linear sized data wrt. the number of nodes. Other scripts will behave differently. Anything that's constant/linear or under a certain threshold (few MBs?) could fall into limited category.

Worth checking https://github.com/openshift/must-gather/tree/master/collection-scripts to see what we actually collect these days. We still don't keep any estimation on which scripts are the biggest contributors.

Reply:
@ingvagabund bear in mind that not all those scripts are gathered by default. A good example is the `gather_audit_logs` (which used to be collected by default in early 4.y versions but not currently, due to huge size of audit logs).

Reply:
> It is important to define what the current behavior is. Looking a year back at the proposal from now will give "the current behavior" a different meaning.

@ingvagabund  I tried to elaborate what actually _current behavior_ means. Let me know what you think.

> Worth making a survey to identify what resources we currently collect and categorizing them based on their importance and impact on solving customer issues/debugging clusters.

That is good idea. What about doing this after setting the foundations around collection-mode in the code. I agree that it would make sense to extend limited mode after the results of the survey.